bdns/dns.go

```diff
@@ -475,3 +488,60 @@ func (dnsClient *DNSClientImpl) LookupCAA(ctx context.Context, hostname string)
 	}
 	return CAAs, nil
 }
+
+// logDNSError logs the provided err result from making a query for hostname. If
+// the err is a `dns.ErrId` instance then the raw Base64 URL encoded bytes of
+// the query (and if not-nil, the response) in wire format is logged as well.
+// This function is called from exchangeOne only for the case where an error
+// occurs querying a hostname that indicates a problem between the VA and the
+// resolver.
+func (dnsClient *DNSClientImpl) logDNSError(hostname string, msg, resp *dns.Msg, err error) {
+	// We use a stand-alone function for this to make it easy to share the
+	// implementation between the DNSClientImpl and MockDNSClient
+	logDNSError(dnsClient.log, hostname, msg, resp, err)
+}
+
+// log a DNS level error with the given logger. See DNSClientImpl.logDNSError.
+func logDNSError(logger blog.Logger, hostname string, msg, resp *dns.Msg, underlying error) {
+	// We don't expect logDNSError to be called with a nil msg or err but
```

Comment:
I'd need to double check, but I think a handful of the network level errors we might see here would return an error from `Exchange` that wouldn't also return a message. The mismatching ID error specifically returns the message because it's more of a sanity check than an actual DNS level error.

Reply:
Oh d'oh, missed that the message here is the one we send, not the response.

Reply:
> I'd need to double check, but I think a handful of the network level errors we might see here would return an error from Exchange that wouldn't also return a message.

The `msg` here is the **input** to `Exchange` which shouldn't ever be empty right? I'm happy to rejig things but I couldn't think of a valid case where this wouldn't be true.